adapters/missena/missena.go

```diff
@@ -0,0 +1,229 @@
+		GDPR:             missenaParams.GDPR,
+		Placement:        missenaParams.Placement,
+		TestMode:         missenaParams.TestMode,
+	}
+	return &missenaRequest
+}
+
+func (a *adapter) makeRequest(missenaParams MissenaInternalParams, reqInfo *adapters.ExtraRequestInfo, imp *openrtb2.Imp, request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+
+	url := a.endpoint + "?t=" + missenaParams.ApiKey
+	parameter := a.makeParameter(missenaParams, request)
+	body, errm := json.Marshal(parameter)
+	if errm != nil {
+		return nil, errm
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		headers.Add("User-Agent", request.Device.UA)
+	}
+
+	if request.Device != nil {
+		if request.Device.IP != "" {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		} else if request.Device.IPv6 != "" {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+	}
+	if request.Site != nil {
+		headers.Add("Referer", request.Site.Page)
+	}
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     url,
+		Headers: headers,
+		Body:    body,
+		ImpIDs:  []string{imp.ID},
+	}, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	// print the request
+
+	var httpRequests []*adapters.RequestData
+	var errors []error
+	gdprApplies, consentString := readGDPR(request)
+
+	missenaInternalParams := MissenaInternalParams{
+		GDPR:        gdprApplies,
+		GDPRConsent: consentString,
+	}
+
+	for _, imp := range request.Imp {
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errors = append(errors, &errortypes.BadInput{
+				Message: "Error parsing bidderExt object",
+			})
+			continue
+		}
+
+		var missenaExt openrtb_ext.ExtImpMissena
+		if err := json.Unmarshal(bidderExt.Bidder, &missenaExt); err != nil {
+			errors = append(errors, &errortypes.BadInput{
+				Message: "Error parsing missenaExt parameters",
+			})
+			continue
+		}
+
+		missenaInternalParams.ApiKey = missenaExt.ApiKey
+		missenaInternalParams.Placement = missenaExt.Placement
+		missenaInternalParams.TestMode = missenaExt.TestMode
+
+		newHttpRequest, err := a.makeRequest(missenaInternalParams, requestInfo, &imp, request)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		httpRequests = append(httpRequests, newHttpRequest)
+
+		break
+	}
+
+	if len(httpRequests) == 0 && len(errors) == 0 {
+		errors = append(errors, &errortypes.BadInput{
+			Message: "No valid impressions found",
+		})
+	}
+
+	return httpRequests, errors
+}
+
+func readGDPR(request *openrtb2.BidRequest) (bool, string) {
+	consentString := ""
+	if request.User != nil {
+		var extUser openrtb_ext.ExtUser
+		if err := json.Unmarshal(request.User.Ext, &extUser); err == nil {
+			consentString = extUser.Consent
+		}
+	}
+	gdprApplies := true
```

Comment:
Just checking, are you sure you always want to default to gdpr applies if no gdpr signal has been provided? Perhaps a supplemental test should be added for when gdpr is not set.